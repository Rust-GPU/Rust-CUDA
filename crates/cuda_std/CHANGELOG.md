# Changelog

Notable changes to this project will be documented in this file.

## Unreleased

- Added warp shuffles, matches, reductions, and votes in the `warp` module.
- Added `activemask` in the `warp` module to query a mask of the active threads.
- Fixed `lane_id` generating invalid ptx.

## 0.2.2 - 2/7/22

- Thread/Block/Grid index/dim intrinsics now hint to llvm that their range is in some bound declared by CUDA. Hopefully allowing for more optimizations.

## 0.2.1 - 12/8/21

- Fixed `shared_array!` not using fully qualified MaybeUninit.
- Fixed `shared_array!` working on the CPU.
- Added experimental dynamic shared memory support through `shared::dynamic_shared_memory`.

## 0.2.0 - 12/5/21

- Added `#[externally_visible]` in conjunction with cg_nvvm dead code elimination changes to mark that
a function is externally visible.
- Added `#[address_space(...)]` in conjunction with cg_nvvm address space changes. Only meant for internal use
and advanced users.
- Added `cuda_std::ptr`.
- Added `is_in_address_space`
- Added `convert_generic_to_specific_address_space`
- Added `convert_specific_address_space_to_generic`
- Added basic static shared memory support with `cuda_std::shared_array`.
