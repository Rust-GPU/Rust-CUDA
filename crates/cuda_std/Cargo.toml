[package]
name = "cuda_std"
version = "0.2.2"
edition = "2018"
license = "MIT OR Apache-2.0"
description = "Standard library for CUDA with rustc_codegen_nvvm"
repository = "https://github.com/Rust-GPU/Rust-CUDA"
readme = "../../README.md"

[dependencies]
vek = { version = "0.17.1", default-features = false, features = ["libm"] }
cuda_std_macros = { version = "0.2", path = "../cuda_std_macros" }
half = "2.4.1"
bitflags = "2.8"
paste = "1.0.15"
