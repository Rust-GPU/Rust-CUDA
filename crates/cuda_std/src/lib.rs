//! # CUDA Standard Library
//!
//! The CUDA Standard Library provides a curated set of abstractions for writing performant, reliable, and
//! understandable GPU kernels using the Rustc NVVM backend.
//!
//! This library will build on non-nvptx targets or targets not using the nvvm backend. However, it will not
//! be usable, and it will throw linker errors if you attempt to use most of the functions in the library.
//! However, [`kernel`] automatically cfg-gates the function annotated for `nvptx64`, therefore,
//! no "actual" functions from this crate should be used when compiling for a non-nvptx target.
//!
//! This crate cannot be used with the llvm ptx backend either, it heavily relies on external functions implicitly
//! defined by the nvvm backend, as well as internal attributes.
//!
//! # Structure
//!
//! This library tries to follow the structure of the Rust standard library to some degree, where
//! different concepts are separated into their own modules.
//!
//! # The Prelude
//!
//! In order to simplify imports, we provide a prelude module which contains GPU analogues to standard library
//! structures as well as common imports such as [`thread`].

#![allow(internal_features)]
#![cfg_attr(
    target_os = "cuda",
    feature(alloc_error_handler, asm_experimental_arch, link_llvm_intrinsics)
)]

extern crate alloc;

pub mod float;
#[allow(warnings)]
pub mod intrinsics;
pub mod io;
pub mod mem;
pub mod misc;
// WIP
// pub mod rt;
pub mod atomic;
pub mod cfg;
pub mod ptr;
pub mod shared;
pub mod thread;
pub mod warp;

mod float_ext;

pub use cuda_std_macros::*;
pub use float::GpuFloat;
pub use float_ext::*;
pub use half;
pub use vek;

pub use half::{bf16, f16};

pub mod prelude {
    pub use crate::f16;
    pub use crate::kernel;
    pub use crate::thread;
    pub use crate::{assert_eq, assert_ne, print, println};
    pub use alloc::{
        borrow::ToOwned,
        boxed::Box,
        format,
        rc::Rc,
        string::{String, ToString},
        vec::Vec,
    };
}

#[cfg(target_arch = "nvptx64")]
#[alloc_error_handler]
fn alloc_handler(layout: core::alloc::Layout) -> ! {
    core::panic!("Memory allocation of {} bytes failed", layout.size());
}

// FIXME(RDambrosio016): For some very odd reason, this function causes an InvalidAddress error when called,
// despite it having no reason for doing that. It needs more debugging to see what is causing it exactly. For now we just trap.
#[cfg(target_arch = "nvptx64")]
#[panic_handler]
fn panic(_info: &core::panic::PanicInfo) -> ! {
    // use crate::prelude::*;
    // let block = thread::block_idx();
    // let thread = thread::thread_idx();

    // let thread_str = if thread.z == 0 && thread.y == 0 {
    //     format!("{}", thread.x)
    // } else if thread.z == 0 {
    //     format!("({}, {})", thread.x, thread.y)
    // } else {
    //     format!("({}, {}, {})", thread.x, thread.y, thread.z)
    // };

    // let block_str = if block.z == 0 && block.y == 0 {
    //     format!("{}", block.x)
    // } else if block.z == 0 {
    //     format!("({}, {})", block.x, block.y)
    // } else {
    //     format!("({}, {}, {})", block.x, block.y, block.z)
    // };

    // let locstr = if let Some(loc) = info.location() {
    //     let file = loc.file().to_string();
    //     let line = loc.line().to_string();
    //     format!("(in file `{}` at line `{}`\n) ", file, line)
    // } else {
    //     String::new()
    // };

    // // let msg = if let Some(s) = info.payload().downcast_ref::<&str>() {
    // //     format!(
    // //         "thread {} in block {} {}panicked: {}",
    // //         thread, block, locstr, s
    // //     )
    // // } else {
    // //     format!("thread {} in block {} {}panicked", thread, block, locstr)
    // // };

    // // crate::println!("{}", msg);

    extern "C" {
        fn __nvvm_trap() -> !;
    }

    unsafe { __nvvm_trap() };
}
