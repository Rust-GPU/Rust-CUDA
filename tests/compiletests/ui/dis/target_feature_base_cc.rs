// build-pass
// only-compute_70
// compile-flags: -Cllvm-args=--disassemble-entry=test_base_cc_70 --error-format=human

// This test verifies feature inheritance for compute_70 (base capability)

use cuda_std::kernel;

#[kernel]
pub unsafe fn test_base_cc_70(result: *mut f32) {
    let mut val = 0.0f32;

    // arch=compute_70 should have target_feature=compute_70 enabled
    #[cfg(target_feature = "compute_70")]
    {
        val += 70.0;
    }

    // arch=compute_70 should also have target_feature=compute_60 enabled (lower capability)
    #[cfg(target_feature = "compute_60")]
    {
        val += 60.0;
    }

    // arch=compute_70 should NOT have target_feature=compute_80 enabled (higher capability)
    #[cfg(target_feature = "compute_80")]
    {
        val += 80.0;
    }

    // arch=compute_70 should NOT have target_feature=compute_90 enabled
    #[cfg(target_feature = "compute_90")]
    {
        val += 90.0;
    }

    // Prevent DCE - expected value should be 130.0 (70 + 60)
    core::ptr::write_volatile(result, val);
}
