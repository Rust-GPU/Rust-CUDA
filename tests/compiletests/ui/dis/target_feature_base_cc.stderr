.visible .entry test_base_cc_70(
    .param .u64 test_base_cc_70_param_0
)
{
    .reg .b32     %r<2>;
    .reg .b64     %rd<3>;
    .loc    1 10 0
$L__func_begin0:
    .loc    1 10 0


    ld.param.u64     %rd1, [test_base_cc_70_param_0];
$L__tmp0:
    .loc    1 10 31
    cvta.to.global.u64     %rd2, %rd1;
$L__tmp1:
    .loc    1 38 5
    bra.uni    $L__tmp2;
$L__tmp2:
    .loc    2 2180 9
    mov.u32     %r1, 1124204544;
    st.volatile.global.u32     [%rd2], %r1;
$L__tmp3:
    .loc    1 39 2
    ret;
$L__tmp4:
$L__func_end0:

}


